neurodsp/filt.py

```diff
@@ -67,204 +66,216 @@ def filter_signal(sig, fs, pass_type, fc, n_cycles=3, n_seconds=None,
-            raise TypeError('butterworth_order must be defined when using an IIR filter.')
-
-    # Process input for FIR filters
+
+def iir_checks(pass_type, n_seconds, butterworth_order, remove_edge_artifacts, verbose):
+    """Checks for using an IIR filter if called from the general filter function."""
+
+    # Check input for IIR filters
+    if remove_edge_artifacts:
+        if verbose:
+            warnings.warn('Edge artifacts are not removed when using an IIR filter.')
+    if pass_type != 'bandstop':
+        if verbose:
+            warnings.warn('IIR filters are not recommended other than for notch filters.')
+    if n_seconds is not None:
+        raise TypeError('n_seconds should not be defined for an IIR filter.')
+    if butterworth_order is None:
+        raise TypeError('butterworth_order must be defined when using an IIR filter.')
+
+
+def iir_filt(sig, pass_type, f_lo, f_hi, f_nyq, butterworth_order, plot_freq_response):
+    """Words, words, words."""
+
+    if pass_type in ('bandpass', 'bandstop'):
+        win = (f_lo / f_nyq, f_hi / f_nyq)
+    elif pass_type == 'highpass':
+        win = f_lo / f_nyq
+    elif pass_type == 'lowpass':
+        win = f_hi / f_nyq
+
+    # Design & apply filter
+    b_vals, a_vals = signal.butter(butterworth_order, win, pass_type)
+    sig_filt = signal.filtfilt(b_vals, a_vals, sig)
+
+    # Plot frequency response, if desired
+    if plot_freq_response:
+        plot_frequency_response(fs, b_vals, a_vals)
+
+    return sig_filt, b_vals, a_vals
+
+
+def fir_filt(sig, fs, pass_type, f_lo, f_hi, f_nyq, n_cycles, n_seconds, plot_freq_response):
+    """Words, words, words."""
+
+    # Compute filter length if specified in seconds
+    if n_seconds is not None:
+        filt_len = int(np.ceil(fs * n_seconds))
     else:
-        # Compute filter length if specified in seconds
-        if n_seconds is not None:
-            filt_len = int(np.ceil(fs * n_seconds))
+        if pass_type == 'lowpass':
+            filt_len = int(np.ceil(fs * n_cycles / f_hi))
         else:
-            if pass_type == 'lowpass':
-                filt_len = int(np.ceil(fs * n_cycles / f_hi))
-            else:
-                filt_len = int(np.ceil(fs * n_cycles / f_lo))
-
-        # Force filter length to be odd
-        if filt_len % 2 == 0:
-            filt_len = int(filt_len + 1)
-
-        # Raise an error if the filter is longer than the signal
-        if filt_len >= len(sig):
-            raise ValueError(
-                """The designed filter (length: {:d}) is longer than the signal (length: {:d}).
-                The filter needs to be shortened by decreasing the n_cycles or n_seconds parameter.
-                However, this will decrease the frequency resolution of the filter.""".format(filt_len, len(sig)))
-
-    # Compute nyquist frequency
-    f_nyq = fs / 2.
+            filt_len = int(np.ceil(fs * n_cycles / f_lo))
+
+    # Force filter length to be odd
+    if filt_len % 2 == 0:
```

Comment:
I'm not seeing a catch notifying the user that the filter order has been changed. If we don't have one, we should add one.

Reply:
I'm not sure this is really needed... the user specifies the filter length in either number of cycles or number of seconds. We convert that to samples, and then then jitter this number by 1 value if it ends up being odd. 

For example, in an alpha filter (8-12 Hz), at default of 3 cycles, and sampling rate of 500 Hz, this trigger does change the filter length from 188 samples to 189 samples, which is the equivalent of changes the filter length from 376 ms to 378 ms. Or, generally, this changes the filter length by 1/fs, where the user has typically specified the filter length in n_cycles, and wouldn't know the exact length in seconds anyway. 

Is this degree of change something we really need to tell the user? A warning would trigger, on average, for every second filter call, and it seems like it might be more confusing/annoying than helpful. 

(I'm pretty sure this is how it all works - someone let me know if I'm missing something here. This changes the filter length, not the filter order).

Reply:
@voytek bump on this small note :)